sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -0,0 +1,622 @@
+
+#include "http/curl/curl.hpp"
+
+#include "azure.hpp"
+#include "http/http.hpp"
+
+#include <string>
+
+using namespace Azure::Core::Http;
+
+std::unique_ptr<Response> CurlTransport::Send(Context& context, Request& request)
+{
+  // Create CurlSession in heap. We will point to it from response's stream to keep it alive
+  CurlSession* session = new CurlSession(request);
+
+  auto performing = session->Perform(context);
+
+  if (performing != CURLE_OK)
+  {
+    switch (performing)
+    {
+      case CURLE_COULDNT_RESOLVE_HOST: {
+        throw Azure::Core::Http::CouldNotResolveHostException();
+      }
+      case CURLE_WRITE_ERROR: {
+        throw Azure::Core::Http::ErrorWhileWrittingResponse();
+      }
+      default: {
+        throw Azure::Core::Http::TransportException();
+      }
+    }
+  }
+
+  auto response = session->GetResponse();
+  // Clean session if no stream was created.
+  // When stream is created, it will delete the session on close
+  if (response->GetBodyStream() == nullptr)
+  {
+    delete session;
+  }
+
+  return response;
+}
+
+CURLcode CurlSession::Perform(Context& context)
+{
+  AZURE_UNREFERENCED_PARAMETER(context);
```

Comment:
(void)context;

Reply:
request

Reply:
please also remove the macro

Reply:
We should keep the macro so that we can standardize everywhere

Reply:
We have an `unused` function defined [here](https://github.com/Azure/azure-sdk-for-cpp/blob/5758bedc5c83c53524c47ea84458ad5d093652c2/sdk/storage/inc/common/storage_common.hpp#L8). ( I can change it to `Unused` if you want.)

It's a C++ style thing, and can be used for multiple variables at one time. For example, `unused(var1, var2, var3);`

C-style macro is always something we want to avoid.

Reply:
@JinmingHu-MSFT, @gilbertw , writing `(void)var;` is the language-native way to indicate that a variable is not used. It works on all compilers and platforms, with the same syntax. I am not sure what we we trying to standardize by using macro, and what the benefits are we getting at all? As a disadvantage, we bring a homemade language extension that nobody needs, and it is a macro, with all the disadvantages you already know. To me, having a macro `#define UNUSED(x) ((void)x);` is the same as `#define INCREMENT(x) ++(x)` - we just invent another syntax to do the thing that already exists in the language, and syntax is the same on all compilers and platforms. Am I missing something here?

Reply:
I think the only problem of `(void)var` is that it's not intuitive. But with `unused`, everybody knows what it's doing.

Reply:
I disagree. As a C and C++ programmer, I know exactly what `(void)var` in the beginning of the function stands for, it is easy, I learned once and I know it forever, it does not take me any time to read it. But it is opposite for the macro - when I as a new developer read the source code in Azure repo, at least for once I need to see what this macro does - to make sure it is doing what I think it is doing (and I expect it to do nothing but `(void)var`). And is it really worth to introduce an additional mental concept, and not only that, but also in a form of macro - for the questionable intuitiveness? (And please excuse me for the choice of words, for some reason they read more attacking than I mean them to be; but in reality I mean to say it in unsarcastically well-wishing tone). I don't think we should be introducing this macro even to improve the language, at least as long as the syntax is the same across all compilers and platforms. Because, it looks like we are starting to implement domain-specific language.  We also could `#define FOREVER while(true)` or `#define FOREVER for(;;)` - so that endless loops look bit more intuitive, but cons outweigh benefits, in my opinion. And also I could understand if it was something we would need to write 3 lies for, and now we need to write one, but it is not the case: it actually takes more characters to use the macro.

BTW, as an alternative, it also could be possible to get away with not naming the function parameter (`void f(int /* unused* /)`) - unless some compiler would complain about naming inconsistency when compiling with an equivalent of `/W4`. But I don't think we should be doing that, I think `(void)param` is slightly better.

Reply:
okay

Reply:
We should use the AZURE_UNREFERENCED_PARAMETER macro.  It makes it very clear why the cast is occurring.  This is a very common macro pattern used in lots of products.  
Additionally it has a benefit that should we ever need to alter the cast behavior/check we can do so in place for all instances.

Reply:
What else could a line `(void)parameter` mean at the top of the function?
What changes do you see theoretically possible to that macro in the future?

Reply:
@BillyONeal, any thoughts?

Reply:
[In the standard library we use `(void)`.](https://github.com/microsoft/STL/blob/0cdf5fbfac83650bf6b404a754be415493466643/stl/inc/deque#L376) Notably, this is the only C-style cast we allow there because it is the only C-style cast which doesn't ever emit code, and the only C-style cast that doesn't cause grep confusion with other things, like function parameter lists. I prefer the cast or the macro solution here because the function solution emits code (which affects debug perf) and because debuggers will step in to it.

Between the cast and the macro I have no strong opinion.

Reply:
leaving as it is for now.
This can be addressed easily on further iterations.  Not really relevant for this PR